src/base/controller.hpp

```diff
@@ -272,39 +292,83 @@ void Controller::clear_parallelization() {
-  if (max_parallel_threads_ < 1)
-  #ifdef _OPENMP
-    max_parallel_threads_ = std::max(1, omp_get_max_threads());
-  #else
-    max_parallel_threads_ = 1;
-  #endif
+  if (max_parallel_experiments_ <= 0)
+    return;
 
-  // Set max_parallel_experiments_
-  parallel_experiments_ = (max_parallel_experiments_ < 1)?
-      std::min<int>({ (int) qobj.circuits.size(), max_parallel_threads_}):
-      std::min<int>({ (int) qobj.circuits.size(), max_parallel_threads_, max_parallel_experiments_});
+  // if memory allows, execute experiments in parallel
+  std::vector<size_t> required_memory_mb_list;
+  for (const Circuit &circ : circuits)
+    required_memory_mb_list.push_back(required_memory_mb(circ));
+  std::sort(required_memory_mb_list.begin(), required_memory_mb_list.end(), std::greater<size_t>());
 
-  int max_num_shots = 0;
-  for (Circuit &circ: qobj.circuits)
-    max_num_shots = std::max<int>({ max_num_shots, (int) circ.shots});
+  int total_memory = 0;
+  parallel_experiments_ = 0;
+  for (int required_memory_mb : required_memory_mb_list) {
+    total_memory += required_memory_mb;
+    if (total_memory > max_memory_mb_)
+      break;
+    ++parallel_experiments_;
+  }
 
-  parallel_shots_ = (max_parallel_shots_ < 1)?
-      std::min<int>({max_num_shots, max_parallel_threads_/parallel_experiments_}):
-      std::min<int>({max_num_shots, max_parallel_threads_/parallel_experiments_, max_parallel_shots_});
-  parallel_shots_ = std::max<int>({1, parallel_shots_});
+  if (parallel_experiments_ == 0) {
+    throw std::runtime_error("a circuit requires more memory than max_memory_mb.");
+  } else if (parallel_experiments_ != 1) {
+    parallel_experiments_ = std::min<int> ({ parallel_experiments_,
+                                             max_parallel_experiments_,
+                                             max_parallel_threads_,
+                                             static_cast<int>(circuits.size()) });
+  }
+}
 
-  parallel_state_update_ = std::max<int>({1, max_parallel_threads_/(parallel_experiments_*parallel_shots_)});
+void Controller::set_parallelization(const Circuit& circ) {
 
-  if(state_available_memory_mb_ > 0)
-  {
-    state_available_memory_mb_ /= parallel_shots_;
-    state_available_memory_mb_ /= parallel_experiments_;
+  if (max_parallel_threads_ < max_parallel_shots_)
+    max_parallel_shots_ = max_parallel_threads_;
+
+  auto circ_memory_mb = required_memory_mb(circ);
+
+  if (max_memory_mb_ < circ_memory_mb)
+    throw std::runtime_error("a circuit requires more memory than max_memory_mb.");
+
+  if (circ_memory_mb == 0)
+    parallel_shots_ = max_parallel_threads_;
+  else
+    parallel_shots_ = std::min<int> ({ static_cast<int>(max_memory_mb_ / circ_memory_mb),
+                                       max_parallel_threads_,
+                                       static_cast<int>(circ.shots) });
+
+  if (max_parallel_shots_ < 1) {
+    // if max_parallel_shots is not configured, nested parallelization is avoided
```

Comment:
The comment refers to the line above, could you move it right above the: `if (max_parallel_shots_ 1) {` ?